### PR TITLE
MDEV-19209: my_getncpus based on threads available

### PR DESCRIPTION
**Pull request by grooverdan**
Detecting the cpus based on sysconf of the online CPUs can significantly
over estimate the number of cpus available.

Whether via numactl, cgroups, taskset, systemd constraints, docker
containers and probably other mechanisms, the number of threads mysqld
can be run on can be quite less.

As such we use the pthread_getaffinity_np function on Linux and FreeBSD
(identical API) to get the number of CPUs.

The number of CPUs is the default for the thread_pool_size and a too
high default will resulting in large memory usage and high context
switching overhead.

I submit this under the MCA.

<pre>
MariaDB [(none)]> show global variables like 'thread_pool_size';
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| thread_pool_size | 8     |
+------------------+-------+
</pre>

restart `numactl -C 0 sql/mysqld ....`

<pre>
MariaDB [(none)]> show global variables like 'thread_pool_size';
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| thread_pool_size | 1     |
+------------------+-------+
</pre>

@Sp1l can you check the FreeBSD bits please. I followed PTHREAD_AFFINITY_NP(3) and CPUSET(9).

### CODE REVIEW

**Comment by grooverdan** [comment]
[MySQL commit](https://github.com/mysql/mysql-server/commit/de9b0a891eb2989d8e2eb7aa6d94c2856b342b6e) seemed happy with FreeBSD and Linux having same implementation.

**Comment by grooverdan** [comment]
Almost a year...... :snail:

**Comment by an3l** [comment]
Pushed to `10.2` with fb01cc3766be9131402b82.
Thanks Daniel.

File: mysys/my_getncpus.c [hunk 1]
```diff
@@ -21,10 +21,36 @@
...
+#else
+      ncpus= 2;
+#endif
+    }
+  }
+
+#else /* __linux__ || FreeBSD && HAVE_PTHREAD_GETAFFINITY_NP */
+
   if (!ncpus)
   {
```
Comment (svoj) [inline mysys/my_getncpus.c]: Why not just add here:
```C
#if (defined(__linux__) || defined(__FreeBSD__)) && defined(HAVE_PTHREAD_GETAFFINITY_NP)
    cpu_set_t set;
    if (pthread_getaffinity_np(pthread_self(), sizeof(set), &set) == 0)
    {
      ncpus= CPU_COUNT(&set);
      return ncpus;
    }
#endif
```

Comment (an3l) [reply mysys/my_getncpus.c]: Make sense.

**Pull request closed without merging**